File: tests/dense_tests/dense_test.py
import numpy as np
from random import randint

# dense matrix multiply vector test 1
# 100x100 random matrix
m1 = np.random.random((100, 100)) * 3
# 100x1 random vector
v1 = np.random.random((100, 1)) * 3
# result of multiplying
r1 = m1.dot(v1)
with open('Dense_matrix_mult_vector_1.txt', 'w') as f:
    # write matrix dimensions in the first line
    f.write(str(np.shape(m1)[0]) + ' ' + str(np.shape(m1)[1]) + '\n')
    # write matrix in the next lines
    for i in m1:
        for j in i:
            f.write(str(j) + ' ')
        f.write('\n')
    # write vector in the next line
    for v in v1:
        for j in v:
            f.write(str(j) + ' ')
    f.write('\n')
    # write result of multiplying in the last line
    for r in r1:
        for j in r:
            f.write(str(j) + ' ')
    f.write('\n')
f.close()


# dense matrix multiply vector test 2
# 30x50 random matrix
m2 = np.random.random((30, 50)) * 5
# 50x1 random vector
v2 = np.random.random((50, 1)) * 5
# result of multiplying
r2 = m2.dot(v2)
with open('Dense_matrix_mult_vector_2.txt', 'w') as f:
    # write matrix dimensions in the first line
    f.write(str(np.shape(m2)[0]) + ' ' + str(np.shape(m2)[1]) + '\n')
    # write matrix in the next lines
    for i in m2:
        for j in i:
            f.write(str(j) + ' ')
        f.write('\n')
    # write vector in the next line
    for v in v2:
        for j in v:
            f.write(str(j) + ' ')
    f.write('\n')
    # write result of multiplying in the last line
    for r in r2:
        for j in r:
            f.write(str(j) + ' ')
    f.write('\n')
f.close()


# dense matrix get N elements test
# 10x10 random matrix
m3 = np.random.random((10, 10)) * 2
# number of elements to get
N = 10
with open('Dense_matrix_get_elems.txt', 'w') as f:
    # write matrix dimensions in the first line
    f.write(str(np.shape(m3)[0]) + ' ' + str(np.shape(m3)[1]) + '\n')
    # write matrix in the next lines
    for i in m3:
        for j in i:
            f.write(str(j) + ' ')
        f.write('\n')
    # write number of elements to get
    f.write(str(N) + '\n')
    # write indexes and value of elements' to get
    for _ in range(N):
        i, j = randint(0, 9), randint(0, 9)
        f.write(str(i) + ' ' + str(j) + ' ' + str(m3[i][j]) + '\n')
f.close()


# dense matrix QR decomposition test
# 5x5 random matrix
m4 = np.random.random((50, 50)) * 10
# get QR decomposition for matrix
q4, r4 = np.linalg.qr(m4)
with open('Dense_matrix_QR_decomposition_1.txt', 'w') as f:
    # write initial matrix dimensions in the first line
    f.write(str(np.shape(m4)[0]) + ' ' + str(np.shape(m4)[1]) + '\n')
    # write initial matrix in the next lines
    for i in m4:
        for j in i:
            f.write(str(j) + ' ')
        f.write('\n')
    # write Q matrix dimensions in the first line
    f.write(str(np.shape(q4)[0]) + ' ' + str(np.shape(q4)[1]) + '\n')
    # write Q matrix in the next lines
    for i in q4:
        for j in i:
            f.write(str(j) + ' ')
        f.write('\n')
    # write R matrix dimensions in the first line
    f.write(str(np.shape(r4)[0]) + ' ' + str(np.shape(r4)[1]) + '\n')
    # write R matrix in the next lines
    for i in r4:
        for j in i:
            f.write(str(j) + ' ')
        f.write('\n')
f.close()

# dense matrix QR decomposition test
# 200x100 random matrix
m5 = np.random.random((200, 100)) * 2
# get QR decomposition for matrix
q5, r5 = np.linalg.qr(m5, mode='complete')
with open('Dense_matrix_QR_decomposition_2.txt', 'w') as f:
    # write initial matrix dimensions in the first line
    f.write(str(np.shape(m5)[0]) + ' ' + str(np.shape(m5)[1]) + '\n')
    # write initial matrix in the next lines
    for i in m5:
        for j in i:
            f.write(str(j) + ' ')
        f.write('\n')
    # write Q matrix dimensions in the first line
    f.write(str(np.shape(q5)[0]) + ' ' + str(np.shape(q5)[1]) + '\n')
    # write Q matrix in the next lines
    for i in q5:
        for j in i:
            f.write(str(j) + ' ')
        f.write('\n')
    # write R matrix dimensions in the first line
    f.write(str(np.shape(r5)[0]) + ' ' + str(np.shape(r5)[1]) + '\n')
    # write R matrix in the next lines
    for i in r5:
        for j in i:
            f.write(str(j) + ' ')
        f.write('\n')
f.close()

File: tests/dense_tests/dense_test.cpp
#include <iostream>
#include <fstream>
#include <cmath>
#include <string>
#include "gtest/gtest.h"
#include "../../src/Dense_matrix.h"

// Dense matrix 100x100 multiply vector test
TEST(Dense_matrix_tests, matrix_mult_vector_1) {
    std::string filename("/home/vljke/Documents/Clion projects/SLAE_4term/tests/dense_tests/Dense_matrix_mult_vector_1.txt");
    std::string temp, s;
    std::ifstream file(filename);
    if (file.is_open()) {
        // Getting matrix dimensions
        std::vector<size_t> MN(2);
        getline(file, temp);
        std::stringstream ssi(temp);
        for (int i = 0; i < 2; ++i) {
            ssi >> s;
            MN[i] = std::stoi(s);
        }
        // Getting matrix from file
        std::vector<std::vector<double>> construct_matrix(MN[0]);
        for (int i = 0; i < MN[0]; ++i) {
            construct_matrix[i].resize(MN[1]);
            getline(file, temp);
            std::stringstream ssm(temp);
            for (int j = 0; j < MN[1]; ++j) {
                ssm >> s;
                construct_matrix[i][j] = std::stod(s);
            }
        }
        Dense_matrix<double> m{construct_matrix};

        // Getting vector to multiply from file
        std::vector<double> v(MN[1]);
        getline(file, temp);
        std::stringstream ssv(temp);
        for (int i = 0; i < MN[1]; ++i) {
            ssv >> s;
            v[i] = std::stod(s);
        }
        // Getting multiply result vector from file
        std::vector<double> r(MN[0]);
        getline(file, temp);
        std::stringstream ssr(temp);
        for (int i = 0; i < MN[1]; ++i) {
            ssr >> s;
            r[i] = std::stod(s);
        }
        // Checking multiply operator
        std::vector<double> res = m * v;
        for (int i = 0; i < MN[0]; ++i)
            ASSERT_NEAR(res[i], r[i], pow(10, -10))
            << "!!! TEST FAILED ON COORDINATE NUMBER " << i << " !!!" << std::endl;
    }
    file.close();
}

// Dense matrix 30x50 multiply vector test
TEST(Dense_matrix_tests, matrix_mult_vector_2) {
    std::string filename("/home/vljke/Documents/Clion projects/SLAE_4term/tests/dense_tests/Dense_matrix_mult_vector_2.txt");
    std::string temp, s;
    std::ifstream file(filename);
    if (file.is_open()) {
        // Getting matrix dimensions
        std::vector<size_t> MN(2);
        getline(file, temp);
        std::stringstream ssi(temp);
        for (int i = 0; i < 2; ++i) {
            ssi >> s;
            MN[i] = std::stoi(s);
        }
        // Getting matrix from file
        std::vector<std::vector<double>> construct_matrix(MN[0]);
        for (int i = 0; i < MN[0]; ++i) {
            construct_matrix[i].resize(MN[1]);
            getline(file, temp);
            std::stringstream ssm(temp);
            for (int j = 0; j < MN[1]; ++j) {
                ssm >> s;
                construct_matrix[i][j] = std::stod(s);
            }
        }
        Dense_matrix<double> m{construct_matrix};

        // Getting vector to multiply from file
        std::vector<double> v(MN[1]);
        getline(file, temp);
        std::stringstream ssv(temp);
        for (int i = 0; i < MN[1]; ++i) {
            ssv >> s;
            v[i] = std::stod(s);
        }
        // Getting multiply result vector from file
        std::vector<double> r(MN[0]);
        getline(file, temp);
        std::stringstream ssr(temp);
        for (int i = 0; i < MN[0]; ++i) {
            ssr >> s;
            r[i] = std::stod(s);
        }
        // Checking multiply operator
        std::vector<double> res = m * v;
        for (int i = 0; i < MN[0]; ++i)
            ASSERT_NEAR(res[i], r[i], pow(10, -10))
            << "!!! TEST FAILED ON COORDINATE NUMBER " << i << " !!!" << std::endl;
    }
    file.close();
}

// Dense matrix 10x10 get N elements test
TEST(Dense_matrix_tests, matrix_get_elements) {
    std::string filename("/home/vljke/Documents/Clion projects/SLAE_4term/tests/dense_tests/Dense_matrix_get_elems.txt");
    std::string temp, s;
    std::ifstream file(filename);
    if (file.is_open()) {
        // Getting matrix dimensions
        std::vector<size_t> MN(2);
        getline(file, temp);
        std::stringstream ssi(temp);
        for (int i = 0; i < 2; ++i) {
            ssi >> s;
            MN[i] = std::stoi(s);
        }
        // Getting matrix from file
        std::vector<std::vector<double>> construct_matrix(MN[0]);
        for (int i = 0; i < MN[0]; ++i) {
            construct_matrix[i].resize(MN[1]);
            getline(file, temp);
            std::stringstream ssm(temp);
            for (int j = 0; j < MN[1]; ++j) {
                ssm >> s;
                construct_matrix[i][j] = std::stod(s);
            }
        }
        Dense_matrix<double> m{construct_matrix};
        // Getting number of elements to get from file
        int N;
        std::vector<double> i_j_val(3);
        getline(file, temp);
        std::stringstream ssn(temp);
        ssn >> s;
        N = std::stoi(s);
        // Getting indexes and values from file
        for (int i = 0; i < N; ++i) {
            getline(file, temp);
            std::stringstream ssj(temp);
            for (int j = 0; j < 3; ++j) {
                ssj >> s;
                i_j_val[j] = std::stod(s);
            }
            ASSERT_NEAR(i_j_val[2], m(i_j_val[0], i_j_val[1]), pow(10, -10))
            << "!!! TEST FAILED ON INDEXES i =  " << i_j_val[0] << ", j = " << i_j_val[0] << " !!!" << std::endl;
        }
    }
    file.close();
}

// Dense matrix 50x50 QR decomposition HH algorithm test
TEST(Dense_matrix_tests, QR_decomp_HH_1) {
    std::string filename("/home/vljke/Documents/Clion projects/SLAE_4term/tests/dense_tests/Dense_matrix_QR_decomposition_1.txt");
    std::string temp, s;
    std::ifstream file(filename);
    if (file.is_open()) {
        // Getting initial matrix dimensions
        std::vector<size_t> MN(2);
        getline(file, temp);
        std::stringstream ssi(temp);
        for (int i = 0; i < 2; ++i) {
            ssi >> s;
            MN[i] = std::stoi(s);
        }
        // Getting initial matrix from file
        std::vector<std::vector<double>> construct_matrix(MN[0]);
        for (int i = 0; i < MN[0]; ++i) {
            construct_matrix[i].resize(MN[1]);
            getline(file, temp);
            std::stringstream ssm(temp);
            for (int j = 0; j < MN[1]; ++j) {
                ssm >> s;
                construct_matrix[i][j] = std::stod(s);
            }
        }
        Dense_matrix<double> m{construct_matrix};

        // Getting Q matrix dimensions
        getline(file, temp);
        std::stringstream ssq(temp);
        for (int i = 0; i < 2; ++i) {
            ssq >> s;
            MN[i] = std::stoi(s);
        }
        // Getting Q matrix from file
        construct_matrix.resize(MN[0]);
        for (int i = 0; i < MN[0]; ++i) {
            construct_matrix[i].resize(MN[1]);
            getline(file, temp);
            std::stringstream ssm(temp);
            for (int j = 0; j < MN[1]; ++j) {
                ssm >> s;
                construct_matrix[i][j] = std::stod(s);
            }
        }
        Dense_matrix<double> Q{construct_matrix};

        // Getting R matrix dimensions
        getline(file, temp);
        std::stringstream ssr(temp);
        for (int i = 0; i < 2; ++i) {
            ssr >> s;
            MN[i] = std::stoi(s);
        }
        // Getting R matrix from file
        construct_matrix.resize(MN[0]);
        for (int i = 0; i < MN[0]; ++i) {
            construct_matrix[i].resize(MN[1]);
            getline(file, temp);
            std::stringstream ssm(temp);
            for (int j = 0; j < MN[1]; ++j) {
                ssm >> s;
                construct_matrix[i][j] = std::stod(s);
            }
        }
        Dense_matrix<double> R{construct_matrix};
        // Testing QR decomposition
        std::pair<Dense_matrix<double>, Dense_matrix<double>> Q_R = m.QR_decomp_HH();
        std::cout << "!!! Start testing QR decomposition !!!" << std::endl;
        // Testing Q matrix first
        std::cout << "Testing Q matrix first:" << std::endl;
        for (int i = 0; i < Q_R.first.getOrder().first; ++i)
            for (int j = 0; j < Q_R.first.getOrder().second; ++j)
                ASSERT_NEAR(Q_R.first(i, j), Q(i, j), pow(10, -10))
                << "!!! TEST FAILED ON MATRIX Q; ELEMENT INDEXES (" << i << ", " << j << ")" << std::endl;
        // Testing R matrix after Q matrix
        std::cout << "Matrix Q is correct, testing R matrix next:" << std::endl;
        for (int i = 0; i < Q_R.second.getOrder().first; ++i)
            for (int j = 0; j < Q_R.second.getOrder().second; ++j)
                ASSERT_NEAR(Q_R.second(i, j), R(i, j), pow(10, -10))
                << "!!! TEST FAILED ON MATRIX R; ELEMENT INDEXES (" << i << ", " << j << ")" << std::endl;
        std::cout << "Matrix R is correct, end of the test." << std::endl;
    }
}

// Dense matrix 200x100 QR decomposition HH algorithm test
TEST(Dense_matrix_tests, QR_decomp_HH_2) {
    std::string filename("/home/vljke/Documents/Clion projects/SLAE_4term/tests/dense_tests/Dense_matrix_QR_decomposition_2.txt");
    std::string temp, s;
    std::ifstream file(filename);
    if (file.is_open()) {
        // Getting initial matrix dimensions
        std::vector<size_t> MN(2);
        getline(file, temp);
        std::stringstream ssi(temp);
        for (int i = 0; i < 2; ++i) {
            ssi >> s;
            MN[i] = std::stoi(s);
        }
        // Getting initial matrix from file
        std::vector<std::vector<double>> construct_matrix(MN[0]);
        for (int i = 0; i < MN[0]; ++i) {
            construct_matrix[i].resize(MN[1]);
            getline(file, temp);
            std::stringstream ssm(temp);
            for (int j = 0; j < MN[1]; ++j) {
                ssm >> s;
                construct_matrix[i][j] = std::stod(s);
            }
        }
        Dense_matrix<double> m{construct_matrix};

        // Getting Q matrix dimensions
        getline(file, temp);
        std::stringstream ssq(temp);
        for (int i = 0; i < 2; ++i) {
            ssq >> s;
            MN[i] = std::stoi(s);
        }
        // Getting Q matrix from file
        construct_matrix.resize(MN[0]);
        for (int i = 0; i < MN[0]; ++i) {
            construct_matrix[i].resize(MN[1]);
            getline(file, temp);
            std::stringstream ssm(temp);
            for (int j = 0; j < MN[1]; ++j) {
                ssm >> s;
                construct_matrix[i][j] = std::stod(s);
            }
        }
        Dense_matrix<double> Q{construct_matrix};

        // Getting R matrix dimensions
        getline(file, temp);
        std::stringstream ssr(temp);
        for (int i = 0; i < 2; ++i) {
            ssr >> s;
            MN[i] = std::stoi(s);
        }
        // Getting R matrix from file
        construct_matrix.resize(MN[0]);
        for (int i = 0; i < MN[0]; ++i) {
            construct_matrix[i].resize(MN[1]);
            getline(file, temp);
            std::stringstream ssm(temp);
            for (int j = 0; j < MN[1]; ++j) {
                ssm >> s;
                construct_matrix[i][j] = std::stod(s);
            }
        }
        Dense_matrix<double> R{construct_matrix};
        // Testing QR decomposition
        std::pair<Dense_matrix<double>, Dense_matrix<double>> Q_R = m.QR_decomp_HH();
        std::cout << "!!! Start testing QR decomposition !!!" << std::endl;
        // Testing Q matrix first
        std::cout << "Testing Q matrix first:" << std::endl;
        for (int i = 0; i < Q_R.first.getOrder().first; ++i)
            for (int j = 0; j < Q_R.first.getOrder().second; ++j)
                ASSERT_NEAR(Q_R.first(i, j), Q(i, j), pow(10, -10))
                << "!!! TEST FAILED ON MATRIX Q; ELEMENT INDEXES (" << i << ", " << j << ")" << std::endl;
        // Testing R matrix after Q matrix
        std::cout << "Matrix Q is correct, testing R matrix next:" << std::endl;
        for (int i = 0; i < Q_R.second.getOrder().first; ++i)
            for (int j = 0; j < Q_R.second.getOrder().second; ++j)
                ASSERT_NEAR(Q_R.second(i, j), R(i, j), pow(10, -10))
                << "!!! TEST FAILED ON MATRIX R; ELEMENT INDEXES (" << i << ", " << j << ")" << std::endl;
        std::cout << "Matrix R is correct, end of the test." << std::endl;
    }
}